File: pcc/py_runtime/src/py_int_shift.c
/* pcc/py_runtime/src/py_int_shift.c
 *
 * Bignum shift helpers split from py_int.c for independent pcc-Python
 * replacement.
 */

#include "py_internal.h"
#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>

static void py_bigint_normalize(PyIntObject *b) {
    while (b->ndigits > 0 && b->digits[b->ndigits - 1] == 0) {
        b->ndigits--;
    }
    if (b->ndigits == 0) b->sign = 0;
    else if (b->sign == 0) b->sign = 1;
}

static PyIntObject *bigint_copy(const PyIntObject *a) {
    PyIntObject *r = py_bigint_alloc(a->ndigits);
    if (r == NULL) return NULL;
    r->sign = a->sign;
    for (int32_t i = 0; i < a->ndigits; i++) r->digits[i] = a->digits[i];
    return r;
}

static PyIntObject *bigint_shl_digits_and_bits(const PyIntObject *a,
                                               uint64_t ndigits_shift,
                                               unsigned bit_shift) {
    if (a->sign == 0) return py_bigint_alloc(0);
    int32_t new_len = (int32_t)(a->ndigits + ndigits_shift + 1);
    PyIntObject *r = py_bigint_alloc(new_len);
    if (r == NULL) return NULL;
    uint32_t carry = 0;
    for (int32_t i = 0; i < a->ndigits; i++) {
        uint64_t cur = ((uint64_t)a->digits[i] << bit_shift) | carry;
        r->digits[i + (int32_t)ndigits_shift] = (uint32_t)(cur & 0xFFFFFFFFu);
        carry = (uint32_t)(cur >> 32);
    }
    r->digits[a->ndigits + (int32_t)ndigits_shift] = carry;
    r->sign = a->sign;
    py_bigint_normalize(r);
    return r;
}

PyIntObject *py_bigint_shl(const PyIntObject *a, uint64_t bits) {
    if (a->sign == 0 || bits == 0) return bigint_copy(a);
    uint64_t nd = bits / 32;
    unsigned nb = (unsigned)(bits % 32);
    if (nb == 0) {
        /* Pure digit shift. */
        int32_t new_len = (int32_t)(a->ndigits + nd);
        PyIntObject *r = py_bigint_alloc(new_len);
        if (r == NULL) return NULL;
        for (int32_t i = 0; i < a->ndigits; i++) {
            r->digits[i + (int32_t)nd] = a->digits[i];
        }
        r->sign = a->sign;
        py_bigint_normalize(r);
        return r;
    }
    return bigint_shl_digits_and_bits(a, nd, nb);
}

PyIntObject *py_bigint_shr(const PyIntObject *a, uint64_t bits) {
    /* Python: negative >> k = floor(a / 2^k). We approximate by doing the
     * magnitude shift, then adjust sign for negatives (round toward -inf).
     * For sign = +1: plain magnitude right shift.
     * For sign = -1: compute q = (|a| + 2^k - 1) >> k; result = -q.
     */
    if (a->sign == 0) return py_bigint_alloc(0);
    uint64_t nd = bits / 32;
    unsigned nb = (unsigned)(bits % 32);

    if ((int64_t)nd >= a->ndigits) {
        if (a->sign < 0) {
            /* floor(negative / big) = -1 (if any bits were nonzero). */
            PyIntObject *r = py_bigint_from_i64(-1);
            return r;
        }
        return py_bigint_alloc(0);
    }

    /* Build |a| shifted. */
    int32_t new_len = a->ndigits - (int32_t)nd;
    PyIntObject *mag = py_bigint_alloc(new_len);
    if (mag == NULL) return NULL;
    for (int32_t i = 0; i < new_len; i++) {
        uint64_t low  = (uint64_t)a->digits[i + (int32_t)nd];
        uint64_t high = (i + (int32_t)nd + 1 < a->ndigits)
                      ? (uint64_t)a->digits[i + (int32_t)nd + 1]
                      : 0;
        uint64_t cur;
        if (nb == 0) cur = low;
        else cur = (low >> nb) | (high << (32 - nb));
        mag->digits[i] = (uint32_t)(cur & 0xFFFFFFFFu);
    }
    mag->sign = 1;
    py_bigint_normalize(mag);

    if (a->sign > 0) return mag;

    /* Negative: check if any shifted-out bits were nonzero. */
    bool tail_nonzero = false;
    if (nb != 0) {
        /* low bits of a->digits[nd] that were chopped */
        uint32_t mask = (1u << nb) - 1u;
        if ((a->digits[nd] & mask) != 0) tail_nonzero = true;
    }
    for (int32_t i = 0; i < (int32_t)nd && !tail_nonzero; i++) {
        if (a->digits[i] != 0) tail_nonzero = true;
    }
    if (tail_nonzero) {
        /* mag += 1 */
        uint32_t carry = 1;
        for (int32_t i = 0; i < mag->ndigits && carry; i++) {
            uint64_t cur = (uint64_t)mag->digits[i] + carry;
            mag->digits[i] = (uint32_t)(cur & 0xFFFFFFFFu);
            carry = (uint32_t)(cur >> 32);
        }
        if (carry) {
            /* grow by one digit */
            PyIntObject *grow = py_bigint_alloc(mag->ndigits + 1);
            if (grow == NULL) { free(mag); return NULL; }
            for (int32_t i = 0; i < mag->ndigits; i++) grow->digits[i] = mag->digits[i];
            grow->digits[mag->ndigits] = carry;
            grow->sign = 1;
            py_bigint_normalize(grow);
            free(mag);
            mag = grow;
        }
    }
    mag->sign = -1;
    py_bigint_normalize(mag);
    return mag;
}

File: pcc/py_runtime/py/py_int_shift.py
"""pcc-Python replacement for py_runtime/src/py_int_shift.c."""
from pcc.extern import extern, c_abi_export, c_int64, c_ptr
from pcc.py_runtime.py.py_abi_constants import (
    PYINTOBJECT_DIGITS_OFFSET,
    PYINTOBJECT_NDIGITS_OFFSET,
    PYINTOBJECT_SIGN_OFFSET,
)
from pcc.unsafe import free, load_i32, ptr_is_null, store_i32


py_bigint_alloc = extern("py_bigint_alloc", (c_int64,), c_ptr)
py_bigint_from_i64 = extern("py_bigint_from_i64", (c_int64,), c_ptr)


def _load_u32(obj, offset: int) -> int:
    v: int = load_i32(obj, offset)
    if v < 0:
        v = v + 4294967296
    return v


def _store_u32(obj, offset: int, value: int) -> None:
    store_i32(obj, offset, value)


def _normalize(b) -> None:
    ndigits: int = load_i32(b, PYINTOBJECT_NDIGITS_OFFSET)
    while ndigits > 0 and _load_u32(b, PYINTOBJECT_DIGITS_OFFSET + (ndigits - 1) * 4) == 0:
        ndigits = ndigits - 1
    store_i32(b, PYINTOBJECT_NDIGITS_OFFSET, ndigits)
    if ndigits == 0:
        store_i32(b, PYINTOBJECT_SIGN_OFFSET, 0)
    elif load_i32(b, PYINTOBJECT_SIGN_OFFSET) == 0:
        store_i32(b, PYINTOBJECT_SIGN_OFFSET, 1)


def _bigint_copy(a):
    ndigits: int = load_i32(a, PYINTOBJECT_NDIGITS_OFFSET)
    r = py_bigint_alloc(ndigits)
    if ptr_is_null(r):
        return r
    store_i32(r, PYINTOBJECT_SIGN_OFFSET, load_i32(a, PYINTOBJECT_SIGN_OFFSET))
    i: int = 0
    while i < ndigits:
        store_i32(r, PYINTOBJECT_DIGITS_OFFSET + i * 4, load_i32(a, PYINTOBJECT_DIGITS_OFFSET + i * 4))
        i = i + 1
    return r


def _shl_digits_and_bits(a, ndigits_shift: int, bit_shift: int):
    if load_i32(a, PYINTOBJECT_SIGN_OFFSET) == 0:
        return py_bigint_alloc(0)
    src_len: int = load_i32(a, PYINTOBJECT_NDIGITS_OFFSET)
    r = py_bigint_alloc(src_len + ndigits_shift + 1)
    if ptr_is_null(r):
        return r
    carry: int = 0
    i: int = 0
    factor: int = 1 << bit_shift
    while i < src_len:
        cur: int = _load_u32(a, PYINTOBJECT_DIGITS_OFFSET + i * 4) * factor + carry
        _store_u32(r, PYINTOBJECT_DIGITS_OFFSET + (i + ndigits_shift) * 4, cur & 4294967295)
        carry = cur >> 32
        i = i + 1
    _store_u32(r, PYINTOBJECT_DIGITS_OFFSET + (src_len + ndigits_shift) * 4, carry)
    store_i32(r, PYINTOBJECT_SIGN_OFFSET, load_i32(a, PYINTOBJECT_SIGN_OFFSET))
    _normalize(r)
    return r


@c_abi_export("py_bigint_shl")
def py_bigint_shl(a, bits: int):
    if load_i32(a, PYINTOBJECT_SIGN_OFFSET) == 0 or bits == 0:
        return _bigint_copy(a)
    nd: int = bits // 32
    nb: int = bits % 32
    src_len: int = load_i32(a, PYINTOBJECT_NDIGITS_OFFSET)
    if nb == 0:
        r = py_bigint_alloc(src_len + nd)
        if ptr_is_null(r):
            return r
        i: int = 0
        while i < src_len:
            store_i32(r, PYINTOBJECT_DIGITS_OFFSET + (i + nd) * 4, load_i32(a, PYINTOBJECT_DIGITS_OFFSET + i * 4))
            i = i + 1
        store_i32(r, PYINTOBJECT_SIGN_OFFSET, load_i32(a, PYINTOBJECT_SIGN_OFFSET))
        _normalize(r)
        return r
    return _shl_digits_and_bits(a, nd, nb)


@c_abi_export("py_bigint_shr")
def py_bigint_shr(a, bits: int):
    if load_i32(a, PYINTOBJECT_SIGN_OFFSET) == 0:
        return py_bigint_alloc(0)
    nd: int = bits // 32
    nb: int = bits % 32
    src_len: int = load_i32(a, PYINTOBJECT_NDIGITS_OFFSET)

    if nd >= src_len:
        if load_i32(a, PYINTOBJECT_SIGN_OFFSET) < 0:
            return py_bigint_from_i64(-1)
        return py_bigint_alloc(0)

    new_len: int = src_len - nd
    mag = py_bigint_alloc(new_len)
    if ptr_is_null(mag):
        return mag

    i: int = 0
    while i < new_len:
        low: int = _load_u32(a, PYINTOBJECT_DIGITS_OFFSET + (i + nd) * 4)
        high: int = 0
        if i + nd + 1 < src_len:
            high = _load_u32(a, PYINTOBJECT_DIGITS_OFFSET + (i + nd + 1) * 4)
        cur: int = low
        if nb != 0:
            cur = (low >> nb) | (high * (1 << (32 - nb)))
        _store_u32(mag, PYINTOBJECT_DIGITS_OFFSET + i * 4, cur & 4294967295)
        i = i + 1
    store_i32(mag, PYINTOBJECT_SIGN_OFFSET, 1)
    _normalize(mag)

    if load_i32(a, PYINTOBJECT_SIGN_OFFSET) > 0:
        return mag

    tail_nonzero: bool = False
    if nb != 0:
        mask: int = (1 << nb) - 1
        if (_load_u32(a, PYINTOBJECT_DIGITS_OFFSET + nd * 4) & mask) != 0:
            tail_nonzero = True
    i = 0
    while i < nd and not tail_nonzero:
        if _load_u32(a, PYINTOBJECT_DIGITS_OFFSET + i * 4) != 0:
            tail_nonzero = True
        i = i + 1

    if tail_nonzero:
        carry: int = 1
        i = 0
        while i < load_i32(mag, PYINTOBJECT_NDIGITS_OFFSET) and carry != 0:
            cur = _load_u32(mag, PYINTOBJECT_DIGITS_OFFSET + i * 4) + carry
            _store_u32(mag, PYINTOBJECT_DIGITS_OFFSET + i * 4, cur & 4294967295)
            carry = cur >> 32
            i = i + 1
        if carry != 0:
            old_len: int = load_i32(mag, PYINTOBJECT_NDIGITS_OFFSET)
            grow = py_bigint_alloc(old_len + 1)
            if ptr_is_null(grow):
                free(mag)
                return grow
            i = 0
            while i < old_len:
                store_i32(grow, PYINTOBJECT_DIGITS_OFFSET + i * 4, load_i32(mag, PYINTOBJECT_DIGITS_OFFSET + i * 4))
                i = i + 1
            _store_u32(grow, PYINTOBJECT_DIGITS_OFFSET + old_len * 4, carry)
            store_i32(grow, PYINTOBJECT_SIGN_OFFSET, 1)
            _normalize(grow)
            free(mag)
            mag = grow

    store_i32(mag, PYINTOBJECT_SIGN_OFFSET, -1)
    _normalize(mag)
    return mag
